pkg/beholder/auth.go

```diff
@@ -1,27 +1,145 @@
+	}
+
+	creds := &authHeaderPerRPCredentials{
+		privKey:   privKey,
+		headerTTL: config.HeaderTTL,
+	}
+	// Initialize the headers ~ lastUpdated is 0 so the headers are generated on the first call
+	creds.refresh()
+	return creds
+}
+
+func (a *authHeaderPerRPCredentials) Credentials() credentials.PerRPCCredentials {
+	return a
+}
+
+func (a *authHeaderPerRPCredentials) GetRequestMetadata(_ context.Context, _ ...string) (map[string]string, error) {
+	return a.getHeaders(), nil
+}
+
+func (a *authHeaderPerRPCredentials) RequireTransportSecurity() bool {
+	return false
+}
+
+// get headers returns the auth headers, refreshing them if they are expired
+func (a *authHeaderPerRPCredentials) getHeaders() map[string]string {
+	if time.Since(a.lastUpdated) > a.headerTTL {
+		a.refresh()
+	}
+	return a.headers
+}
 
-// authHeaderVersion is the version of the auth header format
-var authHeaderVersion = "1"
+// refresh creates a new signed auth header token and sets the lastUpdated time to now
+func (a *authHeaderPerRPCredentials) refresh() {
+	a.lastUpdated = time.Now()
+	switch a.version {
+		// refresh doesn't actually do anything for version 1 since we are only signing the public key
+		// this for backwards compatibility and smooth transition to version 2
+		case authHeaderVersion1:
+			a.headers = BuildAuthHeaders(a.privKey)
+		case authHeaderVersion2:
+			a.headers = buildAuthHeadersV2(a.privKey, &AuthHeaderConfig{timestamp: a.lastUpdated.UnixMilli()})
+		default:
+			a.headers = buildAuthHeadersV2(a.privKey, &AuthHeaderConfig{timestamp: a.lastUpdated.UnixMilli()})
+	}
+	return
+}
 
-// BuildAuthHeaders creates the auth header value to be included on requests.
-// The current format for the header is:
+// AuthHeaderConfig configures buildAuthHeadersV2
+type AuthHeaderConfig struct {
+	timestamp int64
+	version   string
+}
+// BuildAuthHeaders creates the auth headers to be included on requests.
+// There are two formats for the header. Version `1` is:
 //
 // <version>:<public_key_hex>:<signature_hex>
 //
 // where the byte value of <public_key_hex> is what's being signed
+// and <signature_hex> is the signature of the public key.
 func BuildAuthHeaders(privKey ed25519.PrivateKey) map[string]string {
 	pubKey := privKey.Public().(ed25519.PublicKey)
 	messageBytes := pubKey
 	signature := ed25519.Sign(privKey, messageBytes)
-	headerValue := fmt.Sprintf("%s:%x:%x", authHeaderVersion, messageBytes, signature)
 
-	return map[string]string{authHeaderKey: headerValue}
+	return map[string]string{authHeaderKey: fmt.Sprintf("%s:%x:%x", authHeaderVersion1, messageBytes, signature)}
+}
+
+// buildAuthHeadersV2 creates the auth headers to be included on requests.
+// Version `2` is:
+//
+// <version>:<public_key_hex>:<timestamp>:<signature_hex>
+//
+// where the concatenated byte value of <public_key_hex> & <timestamp> is what's being signed
+func buildAuthHeadersV2(privKey ed25519.PrivateKey, config *AuthHeaderConfig) map[string]string {
```

Comment:
Can we get away now with making this function unexported? The exported v1 version is used in https://github.com/smartcontractkit/chainlink/blob/develop/core/services/keystore/beholder.go#L16

Reply:
yes, since it won't be *directly* called by anyone using/configuring a Beholder client. Users should configure a `AuthHeaderProvider` for their client,  which calls `buildAuthHeadersV2` under the hood when auth headers are set via `PerGRPCCredendtials`